### DISCUSSION TITLE
Strange behavior with pinch and scroll

### DISCUSSION THREAD

**Discussion by ac-gl-chris** [AvaloniaUI/Avalonia#12351] [Q&A]
Hi,
I'm trying to get pinch and scroll to work for zoom and pan of i.e. an image.
Here is my problem:

1. Touch display with both fingers and pinch => *PinchEvent* => *Zooming*
2. Release fingers from display => *PinchEndedEvent*
3. Touch display with a single finger and drag => *ScrollGestureEvent* => *Pan*
4. Release that one finger
5. Touch display with a single finger again and drag => **PinchEvent** => **Zooming**

The second time I try to pan the image by using a single finger *PinchEvents* are triggered and hence I'm zooming. From there on every single finger touch and drag on the image are recognized as *PinchEvents*.
When I create a new instance of the UserControl it starts at '1' again.

I tried this with the minimum example as in the *ControlCatalog Gesture Page* (Makes no difference without ScrollViewer as container):
```XML
<ScrollViewer HorizontalScrollBarVisibility="Auto" VerticalScrollBarVisibility="Auto">
    <Image Name="PinchImage" Source="../Assets/checkerboard.png" Stretch="UniformToFill">
	<Image.GestureRecognizers>
		<PinchGestureRecognizer/>
		<ScrollGestureRecognizer ScrollStartDistance="0" CanHorizontallyScroll="True" CanVerticallyScroll="True"/>
	</Image.GestureRecognizers>
    </Image>
</ScrollViewer>
```

I'm running on yocto linux where at least multi touch is working now with xinput.

Best regards
Chris

**Comment by timunie** [comment]
Check out Pan and zoom control https://github.com/wieslawsoltes/PanAndZoom

**Reply by ac-gl-chris** [reply]
From what I see the *ZoomBorder* control from *PanAndZoom* does not support multi touch gestures. There is a *StyledProperty* called *EnableGestureZoom* but it remains unused. I just tried it on the device but it does not zoom using pinch gesture.

**Reply by timunie** [reply]
If you have an idea how to fix it, a PR is welcome. Try it in source under samples -> ControlCatalog

**Comment by ac-gl-chris** [comment]
Hi,

I took the *PinchGestureRecognizer* and added some logging.
It seems that after the first Scroll/Pan using a single finger the *PinchGestureRecognizer* does not receiver the *PointerReleased* after receiving the *PointerPressed* for my single finger. Hence *_firstContact* is not set to null from there on.
For the next touch with a single finger the *PinchGestureRecognizer* receives a *PointerPressed* with a new ID and sets it as *_secondContact* and fires *PinchEvents*.

```
Pinch with two fingers on display:

PinchGestureRecognizer: PointerPressed
    Current States: _firstContact:null _secondContact:null
        New States: _firstContact:1002 _secondContact:null
PinchGestureRecognizer: PointerPressed
    Current States: _firstContact:1002 _secondContact:null
        New States: _firstContact:1002 _secondContact:1003
PinchGestureRecognizer: PointerMoved
    Current States: _firstContact:1002 _secondContact:1003
UserControl with GestureRecognizers: PinchEvent
PinchGestureRecognizer: PointerMoved
    Current States: _firstContact:1002 _secondContact:1003
UserControl with GestureRecognizers: PinchEvent
PinchGestureRecognizer: PointerMoved
    Current States: _firstContact:1002 _secondContact:1003
UserControl with GestureRecognizers: PinchEvent
UserControl with GestureRecognizers: PinchEndedEvent
PinchGestureRecognizer: PointerReleased
    Current States: _firstContact:1002 _secondContact:1003
        New States: _firstContact:1002 _secondContact:null
UserControl with GestureRecognizers: PinchEndedEvent
PinchGestureRecognizer: PointerReleased
    Current States: _firstContact:1002 _secondContact:null
        New States: _firstContact:null _secondContact:null


Then Scroll/Pan with single finger on display:

PinchGestureRecognizer: PointerPressed
    Current States: _firstContact:null _secondContact:null
        New States: _firstContact:1004 _secondContact:null
PinchGestureRecognizer: PointerMoved
    Current States: _firstContact:1004 _secondContact:null
UserControl with GestureRecognizers: ScrollGestureEvent
UserControl with GestureRecognizers: ScrollGestureEvent
UserControl with GestureRecognizers: ScrollGestureEvent
UserControl with GestureRecognizers: ScrollGestureEvent

Then a second Scroll/Pan with a single finger on display:

PinchGestureRecognizer: PointerPressed
    Current States: _firstContact:1004 _secondContact:null
        New States: _firstContact:1004 _secondContact:1005
PinchGestureRecognizer: PointerMoved
    Current States: _firstContact:1004 _secondContact:1005
UserControl with GestureRecognizers: PinchEvent
PinchGestureRecognizer: PointerMoved
    Current States: _firstContact:1004 _secondContact:1005
UserControl with GestureRecognizers: PinchEvent
PinchGestureRecognizer: PointerMoved
    Current States: _firstContact:1004 _secondContact:1005
UserControl with GestureRecognizers: PinchEvent
UserControl with GestureRecognizers: PinchEndedEvent
PinchGestureRecognizer: PointerReleased
    Current States: _firstContact:1004 _secondContact:1005
        New States: _firstContact:1004 _secondContact:null

```

Looks like a bug but I'm not sure if I'm doing something completely wrong =)

Best regards
Chris
